### PR TITLE
soc: nxp: Consolidate nxp soc configs and code

### LINKED ISSUE #76841: i.MX RT1061 not properly supported.
For custom boards using the NXP  i.mx rt1061 there are several build errors. Although the soc is very similar to its extensive sibling the 1062, there are differences. 

- zephyr/soc/nxp/imxrt/imxrt10xx/Kconfig.defconfig: is missing the NUM_IRQS definition for the SOC_MIMXRT1061
- zephyr/soc/nxp/imxrt/imxrt10xx/Kconfig: config SOC_MIMXRT1061 is missing the INIT_ENET_PLL (although it does have two ethernet macs) 
- Probably other definitions are missing, but the first are already blocking the build

### PR DESCRIPTION
**Pull request by decsny**
NXP SOCs seem to be getting a lot of copy pasted code in different places, try to consolidate it, as well as some other minor fixups and restructures.

### CODE REVIEW

File: soc/nxp/kinetis/Kconfig.defconfig [hunk 1]
```diff
@@ -3,14 +3,6 @@
 
 if SOC_FAMILY_KINETIS
 
-config SERIAL_INIT_PRIORITY
-	default 55
-	depends on SERIAL
-
-config ADC_INIT_PRIORITY
```
Comment (mmahadevan108) [inline soc/nxp/kinetis/Kconfig.defconfig:10]: This does not seem to be related to this commit, should we not change this priority?

Comment (decsny) [reply soc/nxp/kinetis/Kconfig.defconfig:10]: it doesn't seem necessary, since it looks like ADC already initializes after DMA

File: soc/nxp/kinetis/Kconfig.defconfig [hunk 2]
```diff
@@ -3,14 +3,6 @@
 
 if SOC_FAMILY_KINETIS
 
-config SERIAL_INIT_PRIORITY
-	default 55
-	depends on SERIAL
```
Comment (dleach02) [inline soc/nxp/kinetis/Kconfig.defconfig:8]: The root of the override of SERIAL is from this PR #39452 to fix a problem that someone found with the order of initialization between SERIAL and the CLOCK control drivers. The root problem was that both classes of drivers had the same priority and by a somewhat undefined linker behavior the initialization order just happened to work until someone built with a different build environment and the linker decided to order things in a reverse order. The referenced PR created the generic config for SERIAL and with a separate commit the NXP parts were forced to this priority.

We would need to make sure that this doesn't get the same initialization priority as the clock control driver before removing this. 

I also note that I had a draft PR that was pulled in favor of the above PR and we appear to have missed pulling over the comment block that was above this code explaining the reasoning for the override.

File: soc/nxp/kinetis/Kconfig.defconfig (hunk 1 above)
Comment (dleach02) [reply soc/nxp/kinetis/Kconfig.defconfig:10]: I suspect this has the same reasoning as the SERIAL.

Comment (dleach02) [reply soc/nxp/kinetis/Kconfig.defconfig:10]: But digging into history and it looks like the initial PR from 2020 set it to this value

File: modules/Kconfig.mcux [hunk 3]
```diff
@@ -366,18 +366,21 @@ config HAS_MCUX_XBARA
 
 config NXP_FW_LOADER
 	bool "Include firmware loader component"
+	default y if (BT || WIFI)
```
Comment (axelnxp) [inline modules/Kconfig.mcux:369]: The firmware loader is not used on all NXP BLE platforms, currently only RW uses it.
MCXW71 loads its firmware through the ROM, using nxp blhost tool.
So, I think we should keep it disabled by default and enable it in the soc defconfig.
Or maybe `default y if SOCXXX` but this could become kinda bloated ?

File: modules/Kconfig.mcux [hunk 4]
```diff
@@ -366,18 +366,21 @@ config HAS_MCUX_XBARA
...
 config NXP_MONOLITHIC_BT
 	bool "BT firmware monolithic build"
+	default y if BT
 	help
 	   If enabled, the BT firmware used by the device will be linked with the
 	   application directly.
 
 config NXP_RF_IMU
 	bool "Include RF_IMU adapter for intercore messaging"
+	default y if (BT || WIFI)
```
Comment (axelnxp) [inline modules/Kconfig.mcux:383]: The RF_IMU component is specific to RW, won't be used on the MCXW family

Comment (decsny) [reply modules/Kconfig.mcux:383]: okay, then should to config be introduced in the RW/Kconfig rather than in this file?

File: soc/nxp/kinetis/Kconfig.defconfig (hunk 2 above)
Comment (decsny) [reply soc/nxp/kinetis/Kconfig.defconfig:8]: the clock control has priority 30 by default, serial has priority 50 by default. so changing it to 55 doesn't seem necessary.

File: soc/nxp/kinetis/Kconfig.defconfig (hunk 1 above)
Comment (decsny) [reply soc/nxp/kinetis/Kconfig.defconfig:10]: updated the commit message

**Comment by decsny** [comment]
* Updated commit message regarding driver priority override removal
* Moved some BT defaults back to RW soc kconfig
* Fixed check compliance errors (except for one)

**Comment by decsny** [comment]
Fixed an accidental squash

**Comment by decsny** [comment]
fixed the last check compliance error (a clang format check error) by disabling clang format around the offending area of code, because clang format was introducing another compliance error when run

@pdgendt looks like you added this check of clang format to check compliance, want to make you aware of this problem now where apparently clang format does not align with the other compliance checks

**Comment by decsny** [comment]
> The clang format checks are informative, and do not block merging the PR

it was blocking CI on the PR until I added this comment to except those lines of code

the reason I pinged you was in case you didn't know there was this conflict between clang format and the other checks

File: soc/nxp/kinetis/Kconfig.defconfig (hunk 2 above)
Comment (dleach02) [reply soc/nxp/kinetis/Kconfig.defconfig:8]: At the time of the change they were in the same priority. I think there is additional activity in the project to break this type of coding of priority and leverage something in device tree to configure dependencies.

Comment (decsny) [reply soc/nxp/kinetis/Kconfig.defconfig:8]: the init priorities are a topic in Arch WG right now, as far as this PR, just want to remove these overrides which are no longer relevant

File: soc/nxp/imxrt/imxrt118x/Kconfig.defconfig [hunk 5]
```diff
@@ -23,12 +23,15 @@ config DCDC_VALUE
 config FLEXSPI_CONFIG_BLOCK_OFFSET
 	default 0x400 if BOOT_FLEXSPI_NOR
 
+config ROM_START_OFFSET
+	default 0xb000 if BOOT_FLEXSPI_NOR && SOC_SERIES_IMXRT118X
```
Comment (danieldegrasse) [inline soc/nxp/imxrt/imxrt118x/Kconfig.defconfig:27]: Isn't this already dependent on `SOC_SERIES_IMXRT118X`, since it is within the `if SOC_SERIES_IMXRT118X` block?

Comment (decsny) [reply soc/nxp/imxrt/imxrt118x/Kconfig.defconfig:27]: yes, it was a copy paste, I'm not sure why you phrased it like a question, I'll just update it to remove it in case I need to update this PR again, otherwise not worth the CI rerun IMO

File: soc/nxp/mcx/Kconfig.defconfig [hunk 6]
```diff
@@ -5,4 +5,32 @@ if SOC_FAMILY_NXP_MCX
...
+choice FLASH_LOG_LEVEL_CHOICE
+	default FLASH_LOG_LEVEL_OFF
+endchoice
+
+choice MEMC_LOG_LEVEL_CHOICE
+	default MEMC_LOG_LEVEL_OFF
+endchoice
+
+# Code relocation is needed when FLASH_MCUX_FLEXSPI_XIP is enabled
+config CODE_DATA_RELOCATION_SRAM
```
Comment (danieldegrasse) [inline soc/nxp/mcx/Kconfig.defconfig]: This seems like it will turn on code relocation whenever `FLASH_MCUX_FLEXSPI_XIP` is enabled, but it looks like the MCX only actually uses code relocation when the FLEXSPI flash driver is enabled: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/mcx/mcxn/CMakeLists.txt#L16, so should this setting depend on `FLASH_MCUX_FLEXSPI_NOR`?

File: drivers/flash/Kconfig [hunk 7]
```diff
@@ -86,10 +86,6 @@ menuconfig FLASH
 
 if FLASH
 
-module = FLASH
```
Comment (danieldegrasse) [inline drivers/flash/Kconfig:89]: To make this change we have to deviate from the pattern present in other driver `Kconfig` files (I looked at the serial and SPI ones for example).

Could we instead disable logging at the SOC level, maybe in `soc/nxp/common/Kconfig.flexspi_xip`?

File: soc/nxp/common/Kconfig.flexspi_xip [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+config FLASH_MCUX_FLEXSPI_XIP
+	bool
+	default $(DT_FLASH_PARENT_IS_FLEXSPI)
+	select XIP
+	help
+	  Allows REfor the soc to safely initialize the clocks for the
+	  FlexSpi when planning to execute code in FlexSpi Memory.
+
+config CODE_DATA_RELOCATION_SRAM
+	default y if FLASH_MCUX_FLEXSPI_XIP && MEMC_MCUX_FLEXSPI
```
Comment (danieldegrasse) [inline soc/nxp/common/Kconfig.flexspi_xip:32]: I *think* the dependency here is on the flash driver, not on the MEMC driver. For example on the RT595 we can use the memc driver while running in XIP mode to interface with HyperRAM on the second FlexSPI device, and we don't need to relocate code to SRAM

File: soc/nxp/common/Kconfig.rom_loader
Comment (danieldegrasse) [inline soc/nxp/common/Kconfig.rom_loader:1]: Commit message needs to be updated, the definition for this Kconfig is not being added to the common flexspi kconfig file

**Review by danieldegrasse — COMMENTED** [review]
Very exciting cleanup, just a few additional comments

**Review by DerekSnell — COMMENTED** [review]
Hi @decsny ,
I don't see an issue with these changes.  However, I want more eyes on this, and don't want my review to count as +1.

Thanks

File: soc/nxp/common/Kconfig.flexspi_xip [hunk 9]
```diff
@@ -0,0 +1,41 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+DT_CHOSEN_Z_FLASH := zephyr,flash
+DT_COMPAT_FLEXSPI := nxp,imx-flexspi
```
[the remark this answers is not in the document]
Comment (decsny) [reply soc/nxp/common/Kconfig.flexspi_xip:5]: thanks but not worried about it right now, i'm sure it can be handled when it's added

File: soc/nxp/rw/Kconfig.defconfig [hunk 10]
```diff
@@ -25,27 +26,15 @@ endif # CORTEX_M_SYSTICK
 
 if BT
 
-config FLASH
+config NXP_MONOLITHIC_BT
```
Comment (EmilioCBen) [inline soc/nxp/rw/Kconfig.defconfig:19]: Where is this used?

File: include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+#endif
+/* clang-format on */
+
+/* Kinetis KE series does not support slew rate. Define macros to have no effect */
+#if defined(CONFIG_SOC_SERIES_KINETIS_KE1XF) || defined(CONFIG_SOC_SERIES_KE1XZ)
+#define PORT_PCR_SRE(x)   0x0
+#define PORT_PCR_SRE_MASK 0x0
+#endif
+
+#define Z_PINCTRL_KINETIS_PINCFG(node_id)                                                          \
```
Comment (manuargue) [inline include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h:47]: Beware that S32K1 is very similar (peripheral wise) to Kinetis KE1 series and both are using the same pinctrl driver. `Z_PINCTRL_KINETIS_PINCFG` is defined `zephyr/soc/nxp/s32/s32k1/pinctrl_soc.h`.

I don't think it's okay to move this header here, because even the hw IP is the same, each soc has an specific integration

**Review by manuargue — DISMISSED** [review]

File: include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h (hunk 11 above)
Comment (dleach02) [reply include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h:47]: @manuargue do you have a recommendation on how this one should be handled? A more generic common name perhaps?

Comment (manuargue) [reply include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h:47]: I'm not requesting a change in the naming, although that's something we will need to do eventually but is a bigger change involving bindings, etc. My request on this particular change is to keep the pinctrl headers for Kinetis under `soc/nxp/kinetis/common/pinctrl_soc.h`. I think the whole point of the `pinctrl_soc.h` header is to be in the soc folder, hence the suffix `_soc` :)

File: drivers/flash/Kconfig (hunk 7 above)
Comment (decsny) [reply drivers/flash/Kconfig:89]: no, I don't care about this pattern, it probably is done thoughtlessly anyways, specific drivers should be able to override driver class config defaults, the other classes just haven't run into the issue yet. I am not putting it in the SOC level because the default value we want has specifically to do with the driver, it belongs here

File: soc/nxp/common/Kconfig.flexspi_xip (hunk 8 above)
Comment (decsny) [reply soc/nxp/common/Kconfig.flexspi_xip:32]: this is what the dependency was before this PR, I just moved it to a different file, while you might be right, I don't want to attempt this change in this PR

File: soc/nxp/mcx/Kconfig.defconfig (hunk 6 above)
Comment (decsny) [reply soc/nxp/mcx/Kconfig.defconfig]: no, it's common to all the SOC due to the drivers being relocated https://github.com/zephyrproject-rtos/zephyr/blob/0a7908588c755d3b8dfa3cc5c967d8a085762500/drivers/flash/CMakeLists.txt#L66

File: soc/nxp/common/Kconfig.rom_loader
Comment (decsny) [reply soc/nxp/common/Kconfig.rom_loader:1]: true, I had originally put it in the other file, I guess I forgot to update the message

File: include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2022, 2024 NXP
```
[the remark this answers is not in the document]
Comment (decsny) [reply include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h:2]: yeah, I just copy pasted the whole file, I suppose I can take this opportunity to delete the (c)

File: include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h (hunk 11 above)
Comment (decsny) [reply include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h:47]: @manuargue completely disagree with this and this is a crucial part of the PR which is important for NPI we are trying to enable which will use this file that is not in the kinetis family, we don't want it in two places. This doesn't even affect s32 if it doesn't choose to include this header in it's pinctrl_soc.h,  kinetis still has it's own pinctrl_soc.h which is just including this file

File: drivers/bluetooth/hci/Kconfig.nxp [hunk 13]
```diff
@@ -25,6 +25,21 @@ config HCI_NXP_SET_CAL_DATA_ANNEX100
 
 if BT_H4_NXP_CTLR
 
+config FLASH
```
Comment (axelnxp) [inline drivers/bluetooth/hci/Kconfig.nxp:28]: I don't think those config should be set under `BT_H4_NXP_CTLR`.
`BT_H4_NXP_CTLR` is not selected on RW, `BT_NXP` is.
`BT_H4_NXP_CTLR` is for host platform that will communicate with a tranceiver through the uart h4 driver.

Lyle from Tony's team should review this too, @dleach02 do you know Lyle's tag on github ?

**Review by axelnxp — CHANGES_REQUESTED** [review]

File: drivers/bluetooth/hci/Kconfig.nxp (hunk 13 above)
Comment (axelnxp) [reply drivers/bluetooth/hci/Kconfig.nxp]: Yes I agree

**Review by axelnxp — APPROVED** [review]
Changes related to BT look good to me, thanks for the clean up

File: drivers/flash/Kconfig (hunk 7 above)
Comment (danieldegrasse) [reply drivers/flash/Kconfig:89]: I think other classes have run into this issue, and the solution was to define a Kconfig for that *driver's* log level, rather than using the subsystem Kconfig: https://github.com/zephyrproject-rtos/zephyr/blob/cfe5bd922432d9c2090e9cbf149edd7172dc4e62/drivers/gpio/Kconfig.fxl6408#L20 Perhaps we should do the same here, because I could see someone wanting debug logging to be enabled for their on chip flash driver, but disabled for the FlexSPI

File: soc/nxp/mcx/Kconfig.defconfig (hunk 6 above)
Comment (danieldegrasse) [reply soc/nxp/mcx/Kconfig.defconfig]: Ok- it looks like you dropped this definition and moved to the common one, which does have the right dependencies (as far as I can tell)

File: drivers/flash/Kconfig (hunk 7 above)
Comment (danieldegrasse) [reply drivers/flash/Kconfig:89]: Looking at this more, some driver Kconfig files define the logging string later (such as I2C: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/Kconfig#L143) and some do not. So I think this approach is fine as is

**Review by danieldegrasse — APPROVED** [review]

**Review by EmilioCBen — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]
